cmd/libs/go2idl/go-to-protobuf/generator.go

```diff
@@ -0,0 +1,488 @@
+					return nil, fmt.Errorf("member %q of %q malformed 'protobuf' tag, tag %d should be key=value, got %q\n", m.Name, t.Name, i+4, extra)
+				}
+				protoExtra[parts[0]] = parts[1]
+			}
+
+			field.Extras = protoExtra
+			if name, ok := protoExtra["name"]; ok {
+				field.Name = name
+				delete(protoExtra, "name")
+			}
+		}
+
+		// extract information from JSON field tag
+		if tag := tags.Get("json"); len(tag) > 0 {
+			parts := strings.Split(tag, ",")
+			if len(field.Name) == 0 && len(parts[0]) != 0 {
+				field.Name = parts[0]
+			}
+			if len(parts) > 1 {
+				for _, s := range parts[1:] {
+					switch s {
+					case "omitempty":
+						// TODO: make this nullable
+						if !field.OptionalSet {
+							field.Optional = true
+							field.OptionalSet = true
+						}
+					case "inline":
+						// TODO: inline all members, give them contextual tag that is non-conflicting
+					}
+				}
+			}
+		}
+
+		if field.Type == nil {
+			if err := memberTypeToProtobufField(&field, m.Type); err != nil {
+				return nil, fmt.Errorf("unable to embed type %q as field %q in %q: %v", m.Type, field.Name, t.Name, err)
+			}
+		}
+		if len(field.Name) == 0 {
+			field.Name = m.Name
+		}
+
+		if field.Map && field.Repeated {
+			// maps cannot be repeated
+			field.Repeated = false
+			// this preserves the distinction between an empty map and a nil map, but might be better implemented another way
+			field.Optional = true
+			field.OptionalSet = true
+		}
+		// embedded fields that are not repeated should be considered required
+		if m.Embedded && !field.Repeated && !field.OptionalSet {
+			field.Optional = false
+			field.OptionalSet = true
+		}
+		if field.Repeated && isPackable(field.Type) {
+			field.Extras["packed"] = "true"
+		}
+		// optional, non-proto3 native types can be embedded
+		if !field.Optional && field.Type.Kind != typesKindProtobuf {
+			field.Extras["(gogoproto.nullable)"] = "false"
+		}
+		fields = append(fields, field)
+	}
+
+	// assign tags
+	highest := 0
+	byTag := make(map[int]*protoField)
+	// fields are in Go struct order, which we preserve
+	for i := range fields {
+		field := &fields[i]
+		tag := field.Tag
+		if tag != -1 {
+			if existing, ok := byTag[tag]; ok {
+				return nil, fmt.Errorf("field %q and %q in %q both have tag %d", field.Name, existing.Name, tag)
+			}
+			byTag[tag] = field
+		}
+		if tag > highest {
+			highest = tag
+		}
+	}
+	// starting from the highest observed tag, assign new field tags
+	for i := range fields {
+		field := &fields[i]
+		if field.Tag != -1 {
+			continue
+		}
+		highest++
+		field.Tag = highest
+		byTag[field.Tag] = field
+	}
+	return fields, nil
+}
+
+func (b bodyGen) doStruct(sw *generator.SnippetWriter) {
+	if len(b.t.Name.Name) == 0 {
+		return
+	}
+	if isPrivateGoName(b.t.Name.Name) {
+		return
+	}
+	fields, err := membersToFields(b.t, b.localPackage)
+	if err != nil {
+		sw.Do(fmt.Sprintf("// ERROR: type $.Name$ cannot be converted to protobuf: %v\n", err), b.t)
```

Comment:
Is it possible to inject non-namer functions into the template?  Didn't
seem obvious how if so.

On Nov 2, 2015, at 4:00 PM, Daniel Smith <notifications@github.com> wrote:

In cmd/libs/go2idl/go-to-protobuf/generator.go
<https://github.com/kubernetes/kubernetes/pull/16649#discussion_r43681338>:

> +		field.Tag = highest
> +		byTag[field.Tag] = field
> +	}
> +	return fields, nil
> +}
> +
> +func (b bodyGen) doStruct(sw *generator.SnippetWriter) {
> +	if len(b.t.Name.Name) == 0 {
> +		return
> +	}
> +	if isPrivateGoName(b.t.Name.Name) {
> +		return
> +	}
> +	fields, err := membersToFields(b.t, b.localPackage)
> +	if err != nil {
> +		sw.Do(fmt.Sprintf("// ERROR: type $.Name$ cannot be converted to protobuf: %v\n", err), b.t)

sw.Do("// ERROR: type $type.Name$ cannot be converted to protobuf:
$.error$\n", err), map[string]interface{}{"type": b.t, "error":
err.Error()})

It's possible to construct a helper function so that the last bit can be
condensed to e.g. args("type", b.t, "error", err.Error())

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16649/files#r43681338>.